types.go

```diff
@@ -1353,6 +1355,85 @@ func (rr *CSYNC) len(off int, compression map[string]struct{}) int {
 	return l
 }
 
+// APL RR. See RFC 3123.
+type APL struct {
+	Hdr      RR_Header
+	Prefixes []APLPrefix `dns:"apl"`
+}
+
+// APLPrefix is an address prefix hold by an APL record.
+type APLPrefix struct {
+	Negation bool
+	Network  net.IPNet
+}
+
+// String returns presentation form of the APL record.
+func (rr *APL) String() string {
+	sb := strings.Builder{}
+	sb.WriteString(rr.Hdr.String())
+	for i, p := range rr.Prefixes {
+		if i != 0 {
+			sb.WriteByte(' ')
+		}
+		sb.WriteString(p.String())
+	}
+	return sb.String()
+}
+
+// String returns presentation form of the APL prefix.
+func (p *APLPrefix) String() string {
+	sb := strings.Builder{}
+	if p.Negation {
+		sb.WriteByte('!')
+	}
+
+	afi := aplAddrLenToFamily[len(p.Network.IP)]
+	sb.WriteString(strconv.Itoa(int(afi)))
+
+	sb.WriteByte(':')
+
+	switch len(p.Network.IP) {
+	case net.IPv4len:
+		sb.WriteString(p.Network.IP.String())
+	case net.IPv6len:
+		// add prefix for IPv4-mapped IPv6
+		if v4 := p.Network.IP.To4(); v4 != nil {
+			sb.WriteString("::ffff:")
```

Comment:
Does `net.IP.String()` not do this already?

Reply:
Unfortunately no. IPv4-mapped IPv6 is as IPv4 per [IP.String](https://godoc.org/net#IP.String).

I'm not sure what are the reasons but Go often stores IPv4 as IPv6 address and then performs conversions in the background so nobody would notice. The conversion to string is one example of that.

APL record is protocol-agnostic and we need to respect the address type and therefore I added the prefix to force the presentation format to be IPv6. I just checked the specification again whether the IPv4-mapped IPv6 is allowed and it is. The other implementation option I considered was to print the address as a pure IPv6 address but that would require different kind of trickery. I would stick to this solution.

Reply:
Ah I see. Don’t love it, but I guess it’s okay. Is there anything wrong with just converting IPv4 as IPv6 address to IPv4 throughout?

Reply:
I think there are the following options to write IPv4 APL in presentation format:

1. `1:192.0.2.0/24`
2. `2:192.0.2.0/24` (invalid)
3. `2:::ffff:192.0.2.0/120`
4. `2:::ffff:c000:0200/120`

Option 1 is plain IPv4. Option 2 is invalid according to the specification, we could be lax and allow it in this library but that would make it incompatible with the other libraries. Options 3 and 4 are equivalent on the wire.

Even though all the valid options are semantically the same (i.e. refer to the same network subnet), the APL has explicit notation of address family so there is an actual difference between 1 and 3 (or 4) in the wire format.

I think we cannot force conversion from IPv4-mapped IPv6 to IPv4 because then part of the information is lost. It would also lead to confusing behavior when converting from write to presentation and back to wire again.

Therefore we should stick to option 3 or 4. Go doesn't have API to print address in that format so we can either write a custom conversion function or use the trick above. For option 4, I also have this if it's prefered:
```go
ip := net.ParseIP("::ffff:192.0.2.1")
fmt.PrintlN(ip) // 192.0.2.1
ip[1] = 1 // hack
fmt.Println(ip.String()[1:]) // ::ffff:c000:201
```